src/main/java/com/google/jenkins/plugins/computeengine/ComputeEngineRetentionStrategy.java

```diff
@@ -0,0 +1,136 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.jenkins.plugins.computeengine;
+
+import com.google.common.collect.ImmutableList;
+import hudson.model.Action;
+import hudson.model.Cause;
+import hudson.model.CauseAction;
+import hudson.model.Executor;
+import hudson.model.ExecutorListener;
+import hudson.model.Job;
+import hudson.model.Queue;
+import hudson.security.ACL;
+import hudson.security.ACLContext;
+import hudson.slaves.RetentionStrategy;
+import java.util.List;
+import java.util.logging.Level;
+import jenkins.model.Jenkins;
+import lombok.extern.java.Log;
+import org.jenkinsci.plugins.durabletask.executors.OnceRetentionStrategy;
+
+/**
+ * A strategy that allows: - setting one shot instances {@link OnceRetentionStrategy} - in case of
+ * preemption of GCP instance to restart preempted tasks
+ */
+@Log
+public class ComputeEngineRetentionStrategy extends RetentionStrategy<ComputeEngineComputer>
+    implements ExecutorListener {
+
+  private final OnceRetentionStrategy delegate;
+  private final boolean oneShot;
+
+  /**
+   * Creates the retention strategy.
+   *
+   * @param retentionTimeMinutes number of minutes of idleness after which to kill the slave; serves
+   *     a backup in case the strategy fails to detect the end of a task
+   * @param oneShot create one shot instance strategy
+   */
+  ComputeEngineRetentionStrategy(int retentionTimeMinutes, boolean oneShot) {
+    this.oneShot = oneShot;
+    delegate = new OnceRetentionStrategy(retentionTimeMinutes);
+  }
+
+  @Override
+  public long check(ComputeEngineComputer c) {
+    return delegate.check(c);
+  }
+
+  @Override
+  public void start(ComputeEngineComputer c) {
+    delegate.start(c);
+  }
+
+  @Override
+  public void taskAccepted(Executor executor, Queue.Task task) {
+    if (oneShot) {
+      delegate.taskAccepted(executor, task);
+    }
+  }
+
+  @Override
+  public void taskCompleted(Executor executor, Queue.Task task, long durationMS) {
+    checkPreempted(executor, task);
+    if (oneShot) {
+      delegate.taskCompleted(executor, task, durationMS);
+    }
+  }
+
+  @Override
+  public void taskCompletedWithProblems(
+      Executor executor, Queue.Task task, long durationMS, Throwable problems) {
+    checkPreempted(executor, task);
+    if (oneShot) {
+      delegate.taskCompletedWithProblems(executor, task, durationMS, problems);
+    }
+  }
+
+  private Queue.Task getBaseTask(Queue.Task task) {
+    Queue.Task parent = task.getOwnerTask();
+    while (task != parent) {
+      task = parent;
+      parent = task.getOwnerTask();
+    }
+    return parent;
+  }
+
+  private void checkPreempted(Executor executor, Queue.Task task) {
```

Comment:
This method combines two orthogonal concerns: 1. checking for prempted 2. re-queueing the task.
Suggest splitting these up into separate methods and calling them from taskCompleted()